### ISSUE TITLE
Disjoint unions on more complex types

### ISSUE THREAD

**Issue by frontendphil** [facebook/flow#6940]
Hey there. We're in the process of introducing flow into our rather large project. I'm currently trying to add the base types. I'm stuck at typing a component that is used to render forms. The problem seems to be that disjoint unions do not work if the types are more complex. 

I've prepared an example [here](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoAlgWwA5wE4AuYASgKYCGAxkVHnJmAOR4XWOqoECe2pYAKj1IBBMAF4wAbwA+YAHblMpAFxNyjMNIC+nIQKEAhcVOkAChUtWMARhu0duvMKIkzUYMADdyMAK4r5X0xrUjwAGncwRwDBXmFUe10nI1dpSO8-AIBnAjx0OQBzCI9o1VjSAwSdJL4AYWNRWUqasAAFOmws40lIgDp+2uKwSJZyABMAeTkYLgB+VWs4OBgKOVRqqF85anQ4OTBa+mwACmwOrNV2uE6ASilI9ChT897o3os+MS+1Rjuejw8LAIvjw+zO1yyvQy-lecAAMnBKD5SABlXL5ArHG6RHQeB5PcGdV5Cd6KT7fGy-e4AsBAkFgl7Q0iwhEIUK1chZUhYnHrDgAHkOOC8Pn8YgARARSDlxVEhGJJJJ5GSrOowFotGBgAA+IA): 

To give you some more context. In our app we're trying to render form fields. Each field consists of a `value` and a `type` prop and some additional props such as `readOnly` etc. What I'm trying to achieve is a connection between the `value` and `type` props. However, both properties can be complex values and therefore also complex types. It seems that this is messing with how disjoint unions work. I've just exact types all along the way which narrowed down the problem but didn't quite solve it. 

I'm thankful for any pointer you may have!

**Comment by frontendphil** [comment]
Maybe related to https://github.com/facebook/flow/issues/6940

**Comment by wchargin** [comment]
You’re correct: Flow doesn’t support propagating constraints in this
way, which is unfortunate. Its disjoint unions are limited to types
where the discriminant can be meaningfully compared with `===`.

See also #6040 (maybe this is the issue that you meant to reference?).
I tried a number of workarounds at the time and didn’t find anything
that worked.

(For what it’s worth, TypeScript doesn’t support this, either.)

**Comment by frontendphil** [comment]
Well shit. So essentially that means that we cannot ensure type-safety for a very large (and important) part of our application. I begin to wonder whether flow and TypeScript have been successfully used in any larger application.

I would understand this restriction when it would apply to "open" types. I understand how comparing them to each other can lead to problems. I don't know the inner working but I would have assumed that 

```es6
type A = {
  foo: 'string'
}

type B = A

// here is what I think flow should do
A === B // true
```

So in other words that types should be comparable like objects and that it should be possible to also resolve more complex disjoint unions.

**Comment by frontendphil** [comment]
The other way round: Do you know whether such an improvement is planned?

**Comment by wchargin** [comment]
> I don't know the inner working but I would have assumed that
>
> ```js
> type A = {
>   foo: 'string'
> }
>
> type B = A
>
> // here is what I think flow should do
> A === B // true
> ```

I don’t understand what you mean. If `type B = A`, then of course Flow
will let you treat any `A` as a `B` and vice versa. But `A === B` is not
a valid expression (or type-expression) if `A` and `B` are types, not
values. Why would you want to compare two types for equality? The
problem described in this issue is an inference/refinement problem, not
a fundamental semantics problem.

To be clear, when you write `if (props.type.name === "A")`, Flow can
successfully refine that `(props.type: TypeA)`. It just doesn’t know how
to infer from this constraint that `props` must be an `A`.

> So essentially that means that we cannot ensure type-safety for a very
> large (and important) part of our application. I begin to wonder
> whether flow and TypeScript have been successfully used in any larger
> application.

I wouldn’t go quite that far. It means that Flow and TypeScript aren’t
equipped to give you _enough_ type safety for a large part of your
application _with its current design_. There are many ways to write
systems that are actually typesafe, but for which proving that they are
typesafe requires mechanisms more complicated than those included with
Flow and TypeScript. Flow in particular aims to support many common
JavaScript patterns, but in the end you _will_ have to take Flow’s
facilities into account when designing your systems, occasionally making
some aesthetic or structural concessions to simplify your program so
that Flow can understand it.

(That said: this particular pattern, of having an object-type
discriminant instead of a string-type discriminant, is quite
benign—you’re not doing anything “dangerously clever”; it’s just that
Flow is unfortunately ignorant.)

> Do you know whether such an improvement is planned?

I have very little visibility into the Flow team’s planning process, but
from what I do know I suspect that no, no such improvement is planned.

**Comment by SamChou19815** [comment]
This is expected limitation
